File: algorithm/binarysearch/makingtteok/makingtteok.py
def makingtteok(start, end, arr, req, l):
    h = (end - start) // 2 + start
    result = 0
    
    for i in arr:
        ih = i - h
        if ih >= 0:
            result += ih
          
    if result >= req: 
        l.append(h)
        if h + 1 >= end: return 1
        makingtteok(h + 1, end, arr, req, l)
    else:
        makingtteok(start, h - 1, arr, req, l)
  

n, m = map(int, input().split())
tteok = list(map(int, input().split()))
maxh = []

makingtteok(0, max(tteok), tteok, m, maxh)
print(max(maxh))

File: algorithm/binarysearch/makingtteok/MakingTteok.java
import java.util.*;

public class MakingTteok {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();
        sc.nextLine();
        int[] tteok = new int[n];
        for (int i = 0; i < n; i++) {
            tteok[i] = sc.nextInt();
        }

        int start = 0;
        int end = Arrays.stream(tteok).max().getAsInt();
        int maxh = 0;
        
        while (start <= end) {
            int result = 0;
            int h = (start + end) / 2;
            for (int i = 0; i < n; i++) {
                if (tteok[i] > h) result += tteok[i] - h;   
            }
            if (result >= m) {
                maxh = h;
                start = h + 1;
            }
            else end = h - 1;
        }
        System.out.println(maxh);
    }
}